From c6fdf15bcafff58dfeb732bfb9bc314cc1d4d4d3 Mon Sep 17 00:00:00 2001
From: Srikanth Krishnakar <Srikanth_Krishnakar@mentor.com>
Date: Mon, 17 Aug 2015 15:58:15 +0530
Subject: [PATCH] ls102xa: LS1021 ARM Generic Timer CompareValue Set 64-bit

Freescale's LS1021 support in u-boot erroneously sets the
64-bit ARM generic PL1 physical time CompareValue register
to all-ones with a 32-bit value.

This causes the timer compare to fire 344 seconds after
u-boot configures it.

URL: https://lists.yoctoproject.org/pipermail/meta-freescale/2015-July/015037.html

Signed-off-by: Srikanth Krishnakar <Srikanth_Krishnakar@mentor.com>
---
 arch/arm/cpu/armv7/ls102xa/timer.c                | 7 ++++---
 arch/arm/include/asm/arch-ls102xa/immap_ls102xa.h | 2 +-
 2 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/arch/arm/cpu/armv7/ls102xa/timer.c b/arch/arm/cpu/armv7/ls102xa/timer.c
index 11b17b2..746bfc0 100644
--- a/arch/arm/cpu/armv7/ls102xa/timer.c
+++ b/arch/arm/cpu/armv7/ls102xa/timer.c
@@ -56,7 +56,8 @@ static inline unsigned long long us_to_tick(unsigned long long usec)
 int timer_init(void)
 {
 	struct sctr_regs *sctr = (struct sctr_regs *)SCTR_BASE_ADDR;
-	unsigned long ctrl, val, freq;
+	unsigned long ctrl, freq;
+	unsigned long long val64;
 
 	/* Enable System Counter */
 	writel(SYS_COUNTER_CTRL_ENABLE, &sctr->cntcr);
@@ -69,8 +70,8 @@ int timer_init(void)
 	asm("mcr p15, 0, %0, c14, c2, 1" : : "r" (ctrl));
 
 	/* Set PL1 Physical Comp Value */
-	val = TIMER_COMP_VAL;
-	asm("mcrr p15, 2, %Q0, %R0, c14" : : "r" (val));
+	val64 = TIMER_COMP_VAL;
+	asm("mcrr p15, 2, %Q0, %R0, c14" : : "r" (val64));
 
 	gd->arch.tbl = 0;
 	gd->arch.tbu = 0;
diff --git a/arch/arm/include/asm/arch-ls102xa/immap_ls102xa.h b/arch/arm/include/asm/arch-ls102xa/immap_ls102xa.h
index edbd5ce..45736db 100644
--- a/arch/arm/include/asm/arch-ls102xa/immap_ls102xa.h
+++ b/arch/arm/include/asm/arch-ls102xa/immap_ls102xa.h
@@ -28,7 +28,7 @@
 #define RCWSR4_SRDS1_PRTCL_SHIFT	24
 #define RCWSR4_SRDS1_PRTCL_MASK		0xff000000
 
-#define TIMER_COMP_VAL			0xffffffff
+#define TIMER_COMP_VAL			0xffffffffffffffffull
 #define ARCH_TIMER_CTRL_ENABLE		(1 << 0)
 #define SYS_COUNTER_CTRL_ENABLE		(1 << 24)
 
-- 
1.9.1

